src/hivpy/demographics.py

```diff
@@ -1,19 +1,149 @@
+                CP[i] = 1.0
+        # make sure CP is exactly 1.0 in case of numerical error
+        CP[N-1] = 1.0
+        return CP
+
+    # Generate Age With Stepwise Distribution
+    # the size of ages should be one larger than the cumulative probabilty
+    # so that each bucket has an upper and lower bound.
+    def genAges(self, N):
+        cpd = self._genCumulativeProb()
+        p0 = 0.0
+        rands = np.random.rand(N)
+        ages_out = np.array([0.0]*N)
+        for i in range(0, cpd.size):
+            # in order to vectorise this method we create a mask of values
+            # that need to change for each age group
+            prob_mask = (p0 < rands) & (rands <= cpd[i])
+            ages_out += ((self.age_boundaries[i]
+                          + (rands - p0)/(cpd[i]-p0)
+                          * (self.age_boundaries[i+1]-self.age_boundaries[i]))
+                         * prob_mask)
+            p0 = cpd[i]
+
+        return ages_out
+
+
+class ContinuousAgeDistribution:
+    """
+    Class to handle age distributions using a continuous probability density.
+    Linear-Exponential function is used for the example distribution.
+
+    y = (mx + c)*exp(A(x-B))
+
+    This fits all three example cases quite neatly.
+    """
+
+    def _integratedLinexp(self, x, m, c, A, B):
+        return np.exp(A*(x-B))*(m*x+c - m/A)/A
+
+    # Example parameters
+    modelParams1 = [-7.19e-4, 5.39e-2, -8.10e-3, 2.12e1]
+    modelParams2 = [-1.03e-3, 7.45e-2, -1.12e-3, 8.47]
+    modelParams3 = [-1.15e-3, 8.47e-2, 2.24e-3, 2.49e1]
+
+    def __init__(self, min_age, max_age, modelParams):
+        self.min_age = min_age
+        model_age_limit = -modelParams[1]/modelParams[0]
+        if(max_age > model_age_limit):
+            warnings.warn("Max age exceeds the maximum age limit for this model (negative probability).\
+                            Adjusting max age to " + model_age_limit)
+            self.max_age = model_age_limit
+        else:
+            self.max_age = max_age
+        self.cpd = lambda x: self._integratedLinexp(x, *modelParams)
+
+    @classmethod
+    def selectModel(cls, inc_cat):
+        if(inc_cat == 1):
+            return cls(-68, 65, cls.modelParams1)
+        elif(inc_cat == 2):
+            return cls(-68, 65, cls.modelParams2)
+        else:
+            return cls(-68, 65, *cls.modelParams3)
+
+    # Generate ages using a (non-normalised) continuous cumulative probability distribution
+    # Given an analytic PD, this should also be analytically defined
+    def genAges(self, N):
+        # Normalise distribution over given range
+        C = self.cpd(self.min_age)
+        M = 1/(self.cpd(self.max_age)-self.cpd(self.min_age))
+
+        def NormDist(x):
+            return M*(self.cpd(x) - C)
+
+        # sample and invert the normalised distribution (in case analytic inverse in impractical)
+        NormX = np.linspace(self.min_age, self.max_age, 11)
```

Comment:
I would use e.g. `num_points = 11` to avoid the magic number here and in the upper limit of `NormY`.